sql/src/main/java/io/crate/metadata/Signature.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata;
+
+import com.google.common.collect.ImmutableList;
+import io.crate.core.collections.Collectors;
+import io.crate.types.AnyType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.UndefinedType;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+public class Signature implements List<DataType> {
+
+    public static final List<Signature> SIGNATURES_SINGLE_NUMERIC = DataTypes.NUMERIC_PRIMITIVE_TYPES.stream()
+        .map(Signature::new)
+        .collect(Collectors.toImmutableList());
+    public static final List<Signature> SIGNATURES_SINGLE_ALL = DataTypes.ALL_TYPES.stream()
+        .map(Signature::new)
+        .collect(Collectors.toImmutableList());
+    public static final List<Signature> SIGNATURES_SINGLE_ANY = ImmutableList.of(new Signature(AnyType.INSTANCE));
+    public static final List<Signature> SIGNATURES_ALL_PAIRS_OF_SAME = DataTypes.ALL_TYPES.stream()
+        .map(dt -> new Signature(dt, dt))
+        .collect(Collectors.toImmutableList());
+
+    /**
+     * Build signatures of all permuted given types including same types e.g. (string, string)
+     */
+    public static List<Signature> permutationsOf(Collection<DataType> dataTypes) {
```

Comment:
This looks a bit confusing as it only permutates 2 types, but what if only 1 or more than 2 are provided?